aws-lc-rs/src/ed25519.rs

```diff
@@ -116,31 +117,22 @@ impl KeyPair for Ed25519KeyPair {
     }
 }
 
-pub(crate) unsafe fn generate_key(rng: &dyn SecureRandom) -> Result<LcPtr<EVP_PKEY>, ()> {
-    // TODO: Should we drop support for using the provided rng in a minor release?
-
-    let mut seed = [0u8; ED25519_SEED_LEN];
-    rng.fill(&mut seed)?;
-
-    let mut public_key = MaybeUninit::<[u8; ED25519_PUBLIC_KEY_LEN]>::uninit();
-    let mut private_key = MaybeUninit::<[u8; ED25519_PRIVATE_KEY_LEN]>::uninit();
-    ED25519_keypair_from_seed(
-        public_key.as_mut_ptr().cast(),
-        private_key.as_mut_ptr().cast(),
-        seed.as_ptr(),
-    );
-    seed.zeroize();
-
-    // ED25519_keypair_from_seed doesn't set FIPS indicator, and Ed25119 is not approved anyways at this time.
-    // Seems like it could be approved for use in the future per FIPS 186-5 and CMVP guidance.
-    crate::fips::set_fips_service_status_unapproved();
-
-    LcPtr::new(EVP_PKEY_new_raw_private_key(
-        EVP_PKEY_ED25519,
-        null_mut(),
-        private_key.assume_init().as_ptr(),
-        ED25519_PRIVATE_KEY_SEED_LEN,
-    ))
+pub(crate) unsafe fn generate_key() -> Result<LcPtr<EVP_PKEY>, ()> {
+    let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, null_mut()) })?;
+
+    if 1 != unsafe { EVP_PKEY_keygen_init(*pkey_ctx) } {
+        return Err(());
+    }
+
+    let mut pkey = null_mut::<EVP_PKEY>();
+
+    if 1 != indicator_check!(unsafe { EVP_PKEY_keygen(*pkey_ctx, &mut pkey) }) {
+        return Err(());
+    }
+
+    let pkey = LcPtr::new(pkey)?;
```

Comment:
How does one produce a key that can be used as an input to the sign function (`pub fn sign(&self, msg: &[u8])`) in this file?

My understanding is that you first have to write the EVP_PKEY object to file in a PKCS format(?).

Reply:
Yeah that is the way the API works currently (an artifact of how the Ring API was designed). The user invokes [`generate_pkcs8`](https://docs.rs/aws-lc-rs/1.3.0/aws_lc_rs/signature/struct.Ed25519KeyPair.html#method.generate_pkcs8) to get a PKCS#8 document back, which they then use [`from_pkcs8`](https://docs.rs/aws-lc-rs/1.3.0/aws_lc_rs/signature/struct.Ed25519KeyPair.html#method.from_pkcs8) to then construct the key. It's a bit silly, and definitely a feature I would like to approve on in the future.

Reply:
would be great if that is possible one day.